### PR TITLE
feat: Pass dashboard context to explore through local storage

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
Currently when user clicks "Edit chart" or chart's title in Dashboard page, we first send a POST request to retrieve form_data_key, then create an Explore URL with that form_data_key and redirect to Explore. We do that in order to pass dashboard context (such as native filters, cross filter, color scheme...) to Explore.
There are several problems with this approach:

- we redirect to another page, so from semantic HTML point of view this element should be a link
- user can't access standard link contextual menu when right clicking
- there's a noticeable delay between a click and redirection because of HTTP request being made in the middle
- we unnecessarily take up space in redis

In order to refactor "Edit chart" to be a link, we need to pass dashboard context in a different way.

This PR implements the following changes:

- save current dashboard state in local storage, where random id generated on Dashboard mount is a key and stringified state is a value
- update the value in local storage on each dashboard state change (such as applying new native filter or cross filter or changing color scheme)
- "Edit chart"/chart title is a React Router Link and it contains local storage key as a search param
- when mounting Explore, retrieve the dashboard state from local storage using the dashboard id search param and merge it with form data coming from initializing Explore request

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/15073128/179526015-608cef82-85fd-432c-aa5b-57f563cd5e0b.mov

### TESTING INSTRUCTIONS
1. Open dashboard
2. Apply native filters, cross filters, filter box filters
3. Open Explore by clicking on chart title or clicking "Edit chart". Try opening in the same tab and opening in a new tab
4. Verify that filters from dashboard are present in Explore

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by kgabryje** [comment]
@jinghua-qa Can you help with testing please?

File: superset-frontend/src/dashboard/actions/hydrate.js
Comment (kgabryje) [inline superset-frontend/src/dashboard/actions/hydrate.js:268]: Some sample dashboards don't have `parents` in their layout metadata, which caused e2e tests to fail

File: superset-frontend/src/dashboard/actions/hydrate.js
Comment (kgabryje) [inline superset-frontend/src/dashboard/actions/hydrate.js:359]: Optimization for cross filters (works the same as the one implemented for native filters a few months ago) to speed up finding charts in scope

File: superset-frontend/src/dashboard/containers/DashboardPage.tsx
Comment (kgabryje) [inline superset-frontend/src/dashboard/containers/DashboardPage.tsx:100]: A new dashboard tab id is generated on each dashboard page opening. We mark ids as redundant when user leaves the dashboard, because they won't be reused. Then we remove redundant dashboard contexts from local storage in order not to clutter it

File: superset-frontend/src/explore/controlUtils/getFormDataWithDashboardContext.ts
Comment (kgabryje) [inline superset-frontend/src/explore/controlUtils/getFormDataWithDashboardContext.ts:112]: This logic is copied from `/superset/utils/core.py:merge_extra_form_data`

File: superset-frontend/src/explore/controlUtils/getFormDataWithDashboardContext.ts
Comment (kgabryje) [inline superset-frontend/src/explore/controlUtils/getFormDataWithDashboardContext.ts:78]: This logic is copied from `/superset/utils/core.py:merge_extra_filters`

**Review by ktmud — COMMENTED** [review]
Thanks for working on this! LGTM in general, just a couple of questions.

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (kgabryje) [reply superset-frontend/src/explore/ExplorePage.tsx:80]: It's `state.nativeFilters.filters` (so without filterSets). We use it in `getFormDataWithExtraFilters` called in line 88. However, now I noticed that we only need `chartsInScope` for each filter - the rest of metadata is redundant. Same goes for cross filters (chartConfiguration). I'll change that

File: superset-frontend/src/dashboard/containers/DashboardPage.tsx
Comment (kgabryje) [reply superset-frontend/src/dashboard/containers/DashboardPage.tsx:144]: Good point!

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx
Comment (kgabryje) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx]: Oops, forgot to remove that. Now exploreUrl is always defined so we don't need a null check.

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (kgabryje) [reply superset-frontend/src/explore/ExplorePage.tsx:104]: 👍

File: superset-frontend/src/dashboard/containers/DashboardPage.tsx
Comment (kgabryje) [reply superset-frontend/src/dashboard/containers/DashboardPage.tsx:164]: That might lead to race conditions between removing local storage key and consuming it in Explore.
Case 1: You open Explore in a new tab and close the tab with dashboard before Explore mounts - key doesn't exist anymore in Explore.
Case 2: You open Explore in the same tab as dashboard - Dashboard component unmounts and removes the key before Explore mounts. That case could be handled by always removing the key in Explore instead of on Dashboard unmount. However, that leads us to Case 3: you open Explore in a new tab, key is removed on Explore mount, then you go back to the tab with dashboard and open a new tab with Explore - key doesn't exist anymore in this tab.

I think flagging key as redundant and removing it on next dashboard action is the safest approach here

File: superset-frontend/src/dashboard/components/SliceHeader/index.tsx
Comment (kgabryje) [reply superset-frontend/src/dashboard/components/SliceHeader/index.tsx:101]: We only need to pass dashboardTabId deep into the component tree, with no intention of ever modifying it or interacting with other redux state variables. I think for such use cases Context is optimal

File: superset-frontend/src/dashboard/containers/DashboardPage.tsx
Comment (kgabryje) [reply superset-frontend/src/dashboard/containers/DashboardPage.tsx:143]: Good idea!

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx
Comment (kgabryje) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:360]: Yeah it does. I'll change that to onClick with history.push

**Comment by michael-s-molina** [comment]
Thanks for the PR @kgabryje. This is a nice improvement now that we have the permalink feature in place and form data keys are not shared anymore.

I noticed that the `form_data_key` is still being generated in Explore. Is there a reason for it? If it's not needed, can we remove it and mark the `explore/form_data` endpoint as deprecated? You can add a [deprecation warning](https://github.com/apache/superset/blob/81bd4968d0a916cb2a20e47b20e31a1434be4f46/superset/views/core.py#L750) to the API file.

**Comment by kgabryje** [comment]
> I noticed that the `form_data_key` is still being generated in Explore. Is there a reason for it? If it's not needed, can we remove it and mark the `explore/form_data` endpoint as deprecated? You can add a [deprecation warning](https://github.com/apache/superset/blob/81bd4968d0a916cb2a20e47b20e31a1434be4f46/superset/views/core.py#L750) to the API file.

The logic related to `form_data_key` is still used in a lot of places in Explore code. We should handle that separately, since this PR focuses on refactoring linking from Dashboard to Explore.

File: superset-frontend/src/dashboard/components/SliceHeader/index.tsx
Comment (ktmud) [reply superset-frontend/src/dashboard/components/SliceHeader/index.tsx]: This all good. I'm not a fan of Redux anyway... just thought creating a new provider has its own overhead---i.e., too many nested context provider always feels wrong.

File: superset-frontend/src/dashboard/containers/DashboardPage.tsx
Comment (kgabryje) [reply superset-frontend/src/dashboard/containers/DashboardPage.tsx]: I ended up creating a new state interface. The state that we save in local storage is trimmed specifically for the use of Explore in order to reduce the size of saved payload (to reduce the risk of overflowing the local storage), so extending interfaces used in Dashboard wasn't applicable

**Review by michael-s-molina — APPROVED** [review]
Code LGTM. Can we get @jinghua-qa's approval before merging the PR? @jinghua-qa It would be a good idea to run our complete test suite on this to make sure everything is working 😉

**Comment by jinghua-qa** [comment]
> Code LGTM. Can we get @jinghua-qa's approval before merging the PR? @jinghua-qa It would be a good idea to run our complete test suite on this to make sure everything is working 😉

OK, i can create a test branch and deploy a workspace to run automation today.

**Comment by jinghua-qa** [comment]
I found one issue that when open the chart from dashboard with color theme, and then save the chart as new chart, the color theme for the new chart still not able to change the color theme. Other than that i think the changes LGTM.

https://user-images.githubusercontent.com/81597121/180700795-d0df7584-0553-47f9-b682-8bdedd4e5571.mov

**Comment by kgabryje** [comment]
Thank you for spotting that @jinghua-qa ! Fixed 

https://user-images.githubusercontent.com/15073128/180782050-86a8a784-cce7-4fd6-a4e9-5a4eea43e81b.mov

**Comment by openaphid** [comment]
I'm afraid it breaks `DASHBOARD_EDIT_CHART_IN_NEW_TAB: True`. Is there a plan to remove this feature flag?

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (john-bodley) [inline superset-frontend/src/explore/ExplorePage.tsx:78]: @kgabryje I believe this logic introduced a regression, i.e., `dataMask` could be `undefined` resulting in an error being thrown [here](https://github.com/apache/superset/blob/e3c6380258a46467ac8e09e64db60aad46bc4655/superset-frontend/src/dashboard/util/activeAllDashboardFilters.ts#L52). Would you be able to look into this issue?

**Pull request merged**
